owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -262,7 +272,12 @@ public void run() {
         uncompletedPieces.put(notStartedIndex, newPieceStatus);
         requestBlockFromPeer(connections.remove(0), newPieceStatus, 0);
       }
-
+      while (!notStartedPieces.isEmpty() && !leecherConnections.isEmpty()) {
+        int notStartedIndex = notStartedPieces.remove();
+        PieceStatus newPieceStatus = makeNewPieceStatus(notStartedIndex);
+        uncompletedPieces.put(notStartedIndex, newPieceStatus);
+        requestBlockFromPeer(leecherConnections.remove(0), newPieceStatus, 0);
```

Comment:
We don't want to request pieces from leechers if 1. They didn't advertise it with a Have message, or 2. They didn't specify they have it in their bitfield.